File: src/skaters/dist.py
"""Gaussian mixture distribution: the distributional prediction type.

A Dist represents a predictive distribution as a weighted mixture of
Gaussians. This is the type that flows through the tree:

    - Leaves emit a Dist (e.g., centered Gaussian with empirical variance)
    - Transforms propagate Dists through their inverses
    - Ensembles combine Dists as weighted mixtures (exact, no approximation)
    - Queries: pdf, cdf, logpdf, quantile, mean, var, std

All math uses only the standard library (math.erf, math.exp, math.log).
No numpy, no scipy. Runs in Pyodide.

A Dist is a list of components: [(weight, mean, std), ...].
Weights are positive and sum to 1. Std is always > 0.
"""

from __future__ import annotations
import math

_SQRT2 = math.sqrt(2.0)
_SQRT2PI = math.sqrt(2.0 * math.pi)
_LOG_SQRT2PI = 0.5 * math.log(2.0 * math.pi)


class Dist:
    """A weighted mixture of Gaussians."""

    __slots__ = ("components",)

    def __init__(self, components: list[tuple[float, float, float]]):
        """Create from a list of (weight, mean, std) tuples.

        Weights must be positive. They will be normalized to sum to 1.
        """
        assert len(components) > 0
        w_total = sum(w for w, _, _ in components)
        assert w_total > 0
        self.components = [(w / w_total, m, s) for w, m, s in components]

    # --- Constructors ---

    @staticmethod
    def gaussian(mean: float = 0.0, std: float = 1.0) -> Dist:
        """A single Gaussian."""
        return Dist([(1.0, mean, std)])

    @staticmethod
    def combine(dists: list[Dist], weights: list[float] | None = None) -> Dist:
        """Weighted mixture of distributions (exact).

        This is the correct way to ensemble distributional predictions.
        """
        n = len(dists)
        if weights is None:
            weights = [1.0 / n] * n
        w_total = sum(weights)
        components = []
        for d, w_outer in zip(dists, weights):
            for w_inner, m, s in d.components:
                components.append((w_outer / w_total * w_inner, m, s))
        return Dist(components)

    # --- Queries ---

    def pdf(self, x: float) -> float:
        """Probability density at x."""
        total = 0.0
        for w, m, s in self.components:
            total += w * _gaussian_pdf(x, m, s)
        return total

    def logpdf(self, x: float) -> float:
        """Log density at x, accumulated in log-space (log-sum-exp).

        A Gaussian mixture is strictly positive everywhere, so for finite x the
        log density is always finite. Computing ``log(sum(w * pdf))`` directly
        underflows to ``-inf`` once x sits many standard deviations from every
        component (e.g. after the scale collapses on a near-constant stream),
        because each ``exp(-z**2 / 2)`` rounds to 0.0. Summing the per-component
        log densities instead keeps the result finite and correct.
        """
        best = -math.inf
        terms = []
        for w, m, s in self.components:
            if w <= 0.0:
                continue
            if s <= 0.0:
                if x == m:
                    return math.inf
                continue
            z = (x - m) / s
            t = math.log(w) - 0.5 * z * z - math.log(s) - _LOG_SQRT2PI
            terms.append(t)
            if t > best:
                best = t
        if best == -math.inf:
            return -math.inf
        return best + math.log(sum(math.exp(t - best) for t in terms))

    def cdf(self, x: float) -> float:
        """Cumulative distribution function at x."""
        total = 0.0
        for w, m, s in self.components:
            total += w * _gaussian_cdf(x, m, s)
        return total

    def crps(self, x: float) -> float:
        """Continuous Ranked Probability Score at observation x.

        CRPS is a proper scoring rule (lower is better) with the same units
        as the data — more robust than logpdf, which is dominated by tails.
        Closed form for a Gaussian mixture (Grimit et al. 2006):

            CRPS = Σ_i w_i A(μ_i - x, σ_i)
                   - ½ Σ_i Σ_j w_i w_j A(μ_i - μ_j, √(σ_i²+σ_j²))

        where A(m, s) = E|N(m, s²)| = m(2Φ(m/s) - 1) + 2s·φ(m/s).
        """
        comps = self.components
        t1 = 0.0
        for w, m, s in comps:
            t1 += w * _abs_expectation(m - x, s)
        t2 = 0.0
        for wi, mi, si in comps:
            for wj, mj, sj in comps:
                t2 += wi * wj * _abs_expectation(mi - mj, math.sqrt(si * si + sj * sj))
        return t1 - 0.5 * t2

    def quantile(self, p: float, tol: float = 1e-9, max_iter: int = 100) -> float:
        """Inverse CDF via bisection."""
        assert 0 < p < 1
        # Initial bracket
        mu = self.mean
        sigma = math.sqrt(self.var)
        lo = mu - 8 * sigma
        hi = mu + 8 * sigma
        for _ in range(max_iter):
            mid = 0.5 * (lo + hi)
            if self.cdf(mid) < p:
                lo = mid
            else:
                hi = mid
            if hi - lo < tol:
                break
        return 0.5 * (lo + hi)

    @property
    def mean(self) -> float:
        """Mean of the mixture."""
        return sum(w * m for w, m, _ in self.components)

    @property
    def var(self) -> float:
        """Variance of the mixture (law of total variance).

        Centered form ``sum w (s^2 + (m - mu)^2)`` rather than
        ``sum w (s^2 + m^2) - mu^2``: the latter catastrophically cancels when
        the means are large relative to the spreads (e.g. a tight component at
        a large mean), silently yielding var=0 and std=0 — an invalid Dist.
        """
        mu = self.mean
        return sum(w * (s * s + (m - mu) * (m - mu)) for w, m, s in self.components)

    @property
    def std(self) -> float:
        """Standard deviation."""
        v = self.var
        return math.sqrt(v) if v > 0 else 0.0

    # --- Transform support ---

    def shift(self, delta: float) -> Dist:
        """Shift all components by delta (for diff inverse)."""
        return Dist([(w, m + delta, s) for w, m, s in self.components])

    def scale(self, factor: float) -> Dist:
        """Scale location and spread by factor (for standardize inverse)."""
        assert factor != 0
        f = abs(factor)
        return Dist([(w, m * factor, s * f) for w, m, s in self.components])

    def affine(self, a: float, b: float) -> Dist:
        """Apply affine transform: x -> a*x + b."""
        assert a != 0
        return Dist([(w, a * m + b, abs(a) * s) for w, m, s in self.components])

    # --- Pruning ---

    def prune(self, max_components: int = 20) -> Dist:
        """Reduce component count by merging closest pairs."""
        max_components = max(1, max_components)   # a Dist needs >=1 component
        if len(self.components) <= max_components:
            return self
        # Sort first so the merge path (and hence the result) is independent of
        # component order — mixtures are order-free but the closest-pair scan
        # is not, and ties at equal means are common (lattice atoms).
        comps = sorted(self.components, key=lambda c: (c[1], c[2], c[0]))
        # Pair selection tolerates last-ulp noise in the means: pick the FIRST
        # pair within a hair of the true minimum distance, so platforms that
        # disagree at the ulp level (e.g. libm erf vs a polynomial) still merge
        # the same pairs in the same order. Exact argmin would amplify ulp
        # noise into macroscopically different mixtures.
        scale = abs(comps[0][1]) + abs(comps[-1][1]) + 1e-12
        while len(comps) > max_components:
            best_dist = float("inf")
            for i in range(len(comps)):
                for j in range(i + 1, len(comps)):
                    d = abs(comps[i][1] - comps[j][1])
                    if d < best_dist:
                        best_dist = d
            thresh = best_dist + 1e-9 * scale
            best_i = best_j = None
            for i in range(len(comps)):
                for j in range(i + 1, len(comps)):
                    if abs(comps[i][1] - comps[j][1]) <= thresh:
                        best_i, best_j = i, j
                        break
                if best_i is not None:
                    break
            if best_i is None:
                # Only reachable when a mean is NaN: every comparison above
                # is False, for the argmin and the threshold scan alike.
                # Merge the first pair so pruning terminates instead of
                # indexing with None.
                best_i, best_j = 0, 1
            # Merge the pair (moment-matching)
            wi, mi, si = comps[best_i]
            wj, mj, sj = comps[best_j]
            w_new = wi + wj
            if w_new < 1e-300:
                # Both weights are effectively zero — just average
                m_new = 0.5 * (mi + mj)
                s_new = max(si, sj, 1e-12)
            else:
                m_new = (wi * mi + wj * mj) / w_new
                # centered form — avoids the same cancellation as Dist.var
                v_new = (wi * (si * si + (mi - m_new) ** 2)
                         + wj * (sj * sj + (mj - m_new) ** 2)) / w_new
                s_new = math.sqrt(max(v_new, 0.0))
            comps[best_i] = (w_new, m_new, s_new)
            comps.pop(best_j)
        return Dist(comps)

    # --- Serialization ---

    def to_dict(self) -> dict:
        return {"components": self.components}

    @staticmethod
    def from_dict(d: dict) -> Dist:
        return Dist([tuple(c) for c in d["components"]])

    # --- Dunder ---

    def __repr__(self) -> str:
        if len(self.components) == 1:
            _, m, s = self.components[0]
            return f"Dist(μ={m:.4g}, σ={s:.4g})"
        return f"Dist({len(self.components)} components, μ={self.mean:.4g}, σ={self.std:.4g})"

    def __len__(self) -> int:
        return len(self.components)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Dist):
            return NotImplemented
        return self.components == other.components


# ---------------------------------------------------------------------------
# Pure-Python Gaussian math
# ---------------------------------------------------------------------------

def _gaussian_pdf(x: float, mean: float, std: float) -> float:
    if std <= 0:
        return float("inf") if x == mean else 0.0
    z = (x - mean) / std
    return math.exp(-0.5 * z * z) / (std * _SQRT2PI)


def _gaussian_cdf(x: float, mean: float, std: float) -> float:
    if std <= 0:
        return 1.0 if x >= mean else 0.0
    return 0.5 * (1.0 + math.erf((x - mean) / (std * _SQRT2)))


def _abs_expectation(m: float, s: float) -> float:
    """E|N(m, s²)| = m(2Φ(m/s) - 1) + 2s·φ(m/s).  (Used by CRPS.)"""
    if s <= 0:
        return abs(m)
    z = m / s
    return m * (2.0 * _gaussian_cdf(z, 0.0, 1.0) - 1.0) + 2.0 * s * _gaussian_pdf(z, 0.0, 1.0)

File: docs/js/skaters/dist.mjs
// Gaussian mixture distribution — JS port of skaters/dist.py.
//
// A Dist is a weighted mixture of Gaussians: components = [[w, m, s], ...].
// Weights are positive and normalized to sum to 1; std is always > 0.
// Pure JS, no dependencies. JS lacks Math.erf, so we provide a
// high-accuracy erf (regularized incomplete gamma) to match Python's
// math.erf closely enough for CDF/quantile parity.

const SQRT2 = Math.SQRT2;
const SQRT2PI = Math.sqrt(2.0 * Math.PI);
const LGAMMA_HALF = 0.5723649429247001; // ln(Gamma(1/2)) = ln(sqrt(pi))

// --- erf via the regularized lower incomplete gamma P(1/2, x^2) ---

function _gser(a, x) {
  // Series expansion for P(a, x), valid for x < a + 1.
  let ap = a;
  let sum = 1.0 / a;
  let del = sum;
  for (let i = 0; i < 300; i++) {
    ap += 1.0;
    del *= x / ap;
    sum += del;
    if (Math.abs(del) < Math.abs(sum) * 1e-17) break;
  }
  return sum * Math.exp(-x + a * Math.log(x) - LGAMMA_HALF);
}

function _gcf(a, x) {
  // Continued fraction for Q(a, x) via the modified Lentz algorithm.
  const TINY = 1e-300;
  let b = x + 1.0 - a;
  let c = 1.0 / TINY;
  let d = 1.0 / b;
  let h = d;
  for (let i = 1; i <= 300; i++) {
    const an = -i * (i - a);
    b += 2.0;
    d = an * d + b;
    if (Math.abs(d) < TINY) d = TINY;
    c = b + an / c;
    if (Math.abs(c) < TINY) c = TINY;
    d = 1.0 / d;
    const del = d * c;
    h *= del;
    if (Math.abs(del - 1.0) < 1e-17) break;
  }
  return Math.exp(-x + a * Math.log(x) - LGAMMA_HALF) * h;
}

function _gammp(a, x) {
  if (x <= 0.0) return 0.0;
  if (x < a + 1.0) return _gser(a, x);
  return 1.0 - _gcf(a, x);
}

export function erf(x) {
  if (x === 0.0) return 0.0;
  const r = _gammp(0.5, x * x);
  return x < 0.0 ? -r : r;
}

// --- pure-JS Gaussian math (mirrors dist.py) ---

function gaussianPdf(x, mean, std) {
  if (std <= 0) return x === mean ? Infinity : 0.0;
  const z = (x - mean) / std;
  return Math.exp(-0.5 * z * z) / (std * SQRT2PI);
}

function gaussianCdf(x, mean, std) {
  if (std <= 0) return x >= mean ? 1.0 : 0.0;
  return 0.5 * (1.0 + erf((x - mean) / (std * SQRT2)));
}

function absExpectation(m, s) {
  // E|N(m, s^2)| = m(2Φ(m/s) - 1) + 2s·φ(m/s).
  if (s <= 0) return Math.abs(m);
  const z = m / s;
  return m * (2.0 * gaussianCdf(z, 0.0, 1.0) - 1.0) + 2.0 * s * gaussianPdf(z, 0.0, 1.0);
}

export class Dist {
  // components: array of [weight, mean, std]
  constructor(components) {
    if (!components || components.length === 0) {
      throw new Error("Dist requires at least one component");
    }
    let wTotal = 0.0;
    for (const c of components) wTotal += c[0];
    if (!(wTotal > 0)) throw new Error("Dist weights must sum to > 0");
    this.components = components.map(([w, m, s]) => [w / wTotal, m, s]);
  }

  // --- constructors ---

  static gaussian(mean = 0.0, std = 1.0) {
    return new Dist([[1.0, mean, std]]);
  }

  static combine(dists, weights = null) {
    const n = dists.length;
    if (weights === null) weights = new Array(n).fill(1.0 / n);
    let wTotal = 0.0;
    for (const w of weights) wTotal += w;
    const components = [];
    for (let i = 0; i < n; i++) {
      const wOuter = weights[i];
      for (const [wInner, m, s] of dists[i].components) {
        components.push([(wOuter / wTotal) * wInner, m, s]);
      }
    }
    return new Dist(components);
  }

  // --- queries ---

  pdf(x) {
    let total = 0.0;
    for (const [w, m, s] of this.components) total += w * gaussianPdf(x, m, s);
    return total;
  }

  logpdf(x) {
    const p = this.pdf(x);
    return p > 0 ? Math.log(p) : -Infinity;
  }

  cdf(x) {
    let total = 0.0;
    for (const [w, m, s] of this.components) total += w * gaussianCdf(x, m, s);
    return total;
  }

  crps(x) {
    // Closed-form CRPS for a Gaussian mixture (Grimit et al. 2006).
    const comps = this.components;
    let t1 = 0.0;
    for (const [w, m, s] of comps) t1 += w * absExpectation(m - x, s);
    let t2 = 0.0;
    for (const [wi, mi, si] of comps) {
      for (const [wj, mj, sj] of comps) {
        t2 += wi * wj * absExpectation(mi - mj, Math.sqrt(si * si + sj * sj));
      }
    }
    return t1 - 0.5 * t2;
  }

  quantile(p, tol = 1e-9, maxIter = 100) {
    if (!(p > 0 && p < 1)) throw new Error("quantile p must be in (0, 1)");
    const mu = this.mean;
    const sigma = Math.sqrt(this.var);
    let lo = mu - 8 * sigma;
    let hi = mu + 8 * sigma;
    for (let i = 0; i < maxIter; i++) {
      const mid = 0.5 * (lo + hi);
      if (this.cdf(mid) < p) lo = mid;
      else hi = mid;
      if (hi - lo < tol) break;
    }
    return 0.5 * (lo + hi);
  }

  get mean() {
    let total = 0.0;
    for (const [w, m] of this.components) total += w * m;
    return total;
  }

  get var() {
    const mu = this.mean;
    let total = 0.0;
    for (const [w, m, s] of this.components) total += w * (s * s + m * m);
    return total - mu * mu;
  }

  get std() {
    const v = this.var;
    return v > 0 ? Math.sqrt(v) : 0.0;
  }

  // --- transform support ---

  shift(delta) {
    return new Dist(this.components.map(([w, m, s]) => [w, m + delta, s]));
  }

  scale(factor) {
    if (factor === 0) throw new Error("scale factor must be nonzero");
    const f = Math.abs(factor);
    return new Dist(this.components.map(([w, m, s]) => [w, m * factor, s * f]));
  }

  affine(a, b) {
    if (a === 0) throw new Error("affine a must be nonzero");
    return new Dist(this.components.map(([w, m, s]) => [w, a * m + b, Math.abs(a) * s]));
  }

  // --- pruning ---

  prune(maxComponents = 20) {
    let comps = this.components.map((c) => c.slice());
    while (comps.length > maxComponents) {
      let bestDist = Infinity;
      let bestI = 0;
      let bestJ = 1;
      for (let i = 0; i < comps.length; i++) {
        for (let j = i + 1; j < comps.length; j++) {
          const d = Math.abs(comps[i][1] - comps[j][1]);
          if (d < bestDist) {
            bestDist = d;
            bestI = i;
            bestJ = j;
          }
        }
      }
      const [wi, mi, si] = comps[bestI];
      const [wj, mj, sj] = comps[bestJ];
      const wNew = wi + wj;
      let mNew, sNew;
      if (wNew < 1e-300) {
        mNew = 0.5 * (mi + mj);
        sNew = Math.max(si, sj, 1e-12);
      } else {
        mNew = (wi * mi + wj * mj) / wNew;
        const vNew =
          (wi * (si * si + mi * mi) + wj * (sj * sj + mj * mj)) / wNew - mNew * mNew;
        sNew = Math.sqrt(Math.max(vNew, 0.0));
      }
      comps[bestI] = [wNew, mNew, sNew];
      comps.splice(bestJ, 1);
    }
    return new Dist(comps);
  }

  // --- serialization ---

  toDict() {
    return { components: this.components.map((c) => c.slice()) };
  }

  static fromDict(d) {
    return new Dist(d.components.map((c) => c.slice()));
  }

  get length() {
    return this.components.length;
  }
}
